packages/block-library/src/heading/block.json

```diff
@@ -2,26 +2,26 @@
 	"name": "core/heading",
 	"category": "common",
 	"attributes": {
-		"align": {
-			"type": "string"
-		},
```

Comment:
This is a breaking change and I'm not certain if we can change the attribute name right now. Imagine templates using it. What happens to templates...

Reply:
`migrate` doesn't work with templates which is pretty unfortunate as it means that all this logic used with `deprecated` is totally unreliable with block definitions :(

Reply:
Seems like we could solve this thought? Try validating and migrating?

Reply:
We could try something along those lines:

```diff
diff --git a/lib/compat.php b/lib/compat.php
index fb09a6e49..6e0c5947b 100644
--- a/lib/compat.php
+++ b/lib/compat.php
@@ -24,3 +24,25 @@ function gutenberg_safe_style_css_column_flex_basis( $attr ) {
 	return $attr;
 }
 add_filter( 'safe_style_css', 'gutenberg_safe_style_css_column_flex_basis' );
+
+function myplugin_register_book_post_type() {
+	$args = array(
+		'public' => true,
+		'label'  => 'Books',
+		'show_in_rest' => true,
+		'template' => array(
+			array( 'core/image', array(
+				'align' => 'left',
+			) ),
+			array( 'core/heading', array(
+				'placeholder' => 'Add Author...',
+				'align' => 'right',
+			) ),
+			array( 'core/paragraph', array(
+				'placeholder' => 'Add Description...',
+			) ),
+		),
+	);
+	register_post_type( 'book', $args );
+}
+add_action( 'init', 'myplugin_register_book_post_type' );
diff --git a/packages/blocks/src/api/factory.js b/packages/blocks/src/api/factory.js
index 304329d0b..84ed5598d 100644
--- a/packages/blocks/src/api/factory.js
+++ b/packages/blocks/src/api/factory.js
@@ -41,6 +41,15 @@ export function createBlock( name, attributes = {}, innerBlocks = [] ) {
 	// Get the type definition associated with a registered block.
 	const blockType = getBlockType( name );
 
+	if ( blockType.deprecated ) {
+		blockType.deprecated.forEach( ( { migrate } ) => {
+			if ( ! migrate || every( attributes, ( _, key ) => blockType.attributes[ key ] ) ) {
+				return;
+			}
+			( [ attributes ] = castArray( migrate( attributes, innerBlocks ) ) );
+		} );
+	}
+
 	// Ensure attributes contains only values defined by block type, and merge
 	// default values for missing attributes.
 	const sanitizedAttributes = reduce( blockType.attributes, ( result, schema, key ) => {
```

However, it might not work as expected as it's hard to tell when an attribute is properly migrated. It doesn't cover nested blocks at all. 

I'm tempted to leave the name of attribute unchanged to stay on the safe side.

Reply:
Done in 8a464f4